### ISSUE TITLE
[Question] Query by dbref field

### ISSUE THREAD

**Issue by anatoly-kryzhanovsky** [litedb-org/LiteDB#1186]
I have small question about using dbref: for example i have following model:
```
class Order
{
  [BsonId]
  public Guid Id {get;set;}

  [BsonRef]
  public Customer Customer {get;set;}
  public DateTime OrderDate {get;set;}
}

class Customer
{
  [BsonId]
  public Guid Id {get;set;}

  public string Name {get;set;}
}
```

and i want to obtain all orders where customer name contains 'Bob'
```
var db = new LiteDatabase(path);
var orders = db
  .GetCollection<Order>()
  .IncludeAll()
  .Find(Query.Where(NameField, x => x.AsString.Contains('Bob'))
  .ToList()
```

but collections always empty. I try to setup logger instance and gain any additional information (may be error), but can't find any usefull info.
i try different values for const NameField:
```
$.Customer.Name
Customer.Name
Name
```

but no luck

i try to do that with IQueryable syntax:
```
var orders = db
  .GetCollection<Order>()
  .IncludeAll()
  .Find(x => x.Customer.Name.Contains('Bob'))
  .ToList()
```

with same result - empty collection.

so question is simple - can i use properties of dbref field in query or not.

- information about environment:
- lite db version: 4.1.4
- xamarin android sdk - 9.1.7.0

**Comment by edcararo** [comment]
was anybody able to solve this problem? thanks

**Comment by d2funlife** [comment]
As option

```
var targetUsers = db.GetCollection<Customer>()
                    .Find(x => x.Name.Contains("Bob"))
                    .ToList();

var targetUsersIds = targetUsers.Select(x => x.Id);


var orders = db.GetCollection<Order>()
               .Include(x => x.Customer)
               .Find(x => targetUsersIds.Contains(x.Customer.Id))
               .ToList();
```

**Comment by arrteamgun** [comment]
@mbdavid 
Can we please have some kind of simple solution for that in 5.0.0?
 
We have pretty the same issue
When you have a chain of instances like:
A -> B -> C -> D (-> is a foreign key)
We'd like to get some A instances using foreign key and D instance in condition. But we can't even use C instance. It's a crucial moment I beleive.

**Comment by mbdavid** [comment]
In v5 you can query as expected. v5 contains a query engine that will made include before run filter. You can try in SQL like:

```
EXPLAIN
SELECT $
FROM Order
INCLUDE Customer
WHERE Customer.Name like 'JOHN%'
```

(using EXPAIN you will see how engine will resolve your query - remove EXPLAIN to run)

**Comment by JKamsker** [comment]
Yes, in LiteDB 5. In v4 the include ran after the filter, which is why you always got an empty result. v5 resolves the reference first whenever the filter uses it, as mbdavid said above. Checked on 5.0.21:

```csharp
var orders = db.GetCollection<Order>()
    .Include(x => x.Customer)
    .Find(x => x.Customer.Name.Contains("Bob"))
    .ToList();
```

The SQL equivalent is `SELECT $ FROM Order INCLUDE Customer WHERE Customer.Name LIKE '%Bob%'`. The `Include` is required here. Without it, the stored reference only contains `$id`/`$ref`, and nothing matches.

Filtering by the customer's Id needs no `Include`: `x => x.Customer.Id == id` maps to `$.Customer.$id`. Filters on other referenced fields load the reference for every order and can't use an index. On large collections, select the matching customer ids first and use `ids.Contains(x.Customer.Id)`, as d2funlife showed. #1888 has more details.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
